python-package/lightgbm/basic.py

```diff
@@ -2407,6 +2407,13 @@ def predict(self, data, num_iteration=None,
         result : numpy array
             Prediction result.
         """
+        params_to_update = copy.deepcopy(self.params)
+        params_to_update.update(dict(kwargs,
+                                     predict_raw_score=raw_score,
+                                     predict_leaf_index=pred_leaf,
+                                     predict_contrib=pred_contrib,
+                                     num_iteration_predict=num_iteration))
+        self.reset_parameter(params_to_update)
```

Comment:
@guolinke This seems to be quite computationally expensive part and it doesn't work in case params contain any "core" parameters, e.g.
```
[LightGBM] [Fatal] Cannot change metric during training
[LightGBM] [Fatal] Cannot change num_class during training
[LightGBM] [Fatal] Cannot change boosting during training
```

Maybe we can put predict parameters into config directly at cpp side during prediction itself?

Reply:
Yeah, use reset parameter is not a good idea. BTW, do we need to save predict parameters?

Reply:
Do you mean remove predict params from model file completely? I think it makes sense, because they are not needed to restore a model and seems to be more "logging" stuff (or at least something different than params used to train model). I think we can introduce naming schema like `predict_*` for them and filter later to not write params in model file which start with `predict_`.

Reply:
yeah, it makes senses.

Reply:
OK, then I'll remove my attempts to record prediction time params from here. Can you please help with cpp part to not store them at all (here or in a separate PR)?

Reply:
I think it is generated automatically: https://github.com/microsoft/LightGBM/blob/master/src/io/config_auto.cpp#L660-L663. 
Maybe we can add a tag, e.g. `[no-save]`, to these parameters, and skip their `to_string`.

Reply:
Cool idea!

Reply:
Hmmm, seems that we already have `[doc-only]` directive:
https://github.com/microsoft/LightGBM/blob/dc65e0ac6ee8faa5790263a8dc2d9cc2a8dd7c4b/helpers/parameter_generator.py#L301-L302

BTW, why are for example `boosting` and `objective` already `doc-only`?

Reply:
As they are not automatically generated. We manually write their code.

Reply:
@guolinke 
> Maybe we can add a tag, e.g. [no-save], to these parameters, and skip them in to_string.

I've added `[no-save]` tag in the latest commit and applied it to `predict` and `convert_model` tasks' params.
Maybe we can apply it to "output" training values as well? I mean, params like
https://github.com/microsoft/LightGBM/blob/102893a11c1dd712bb70d87b3b129ddf42daf06f/include/LightGBM/config.h#L488-L491
https://github.com/microsoft/LightGBM/blob/102893a11c1dd712bb70d87b3b129ddf42daf06f/include/LightGBM/config.h#L564-L567
https://github.com/microsoft/LightGBM/blob/102893a11c1dd712bb70d87b3b129ddf42daf06f/include/LightGBM/config.h#L784-L787

Reply:
@guolinke WDYT about the list of candidate params for the tag?